### PR TITLE
Intersection search

### PR DESCRIPTION
**Pull request by aarthykc**
### Context
Adding unit tests for intersection search and ability to  index and then search for an intersection. 

### Explanation 
The problem that we currently have with carmen especially with addresses is that we index a lot of features that have the same relevance and score and because of the limits we have in place for verifymatch and spatialmatch, a specific feature we'd like to surface, just by virtue of being lower down in the sort id filtered out. 
While thinking of how to solve this problem I started looking at how we index house numbers. We don’t create a new feature for every house number: we have one feature and an `addressnumber` array with all the house numbers. Essentially, for intersection search we'll want to take a similar approach, i.e. we want fewer features which more densely populated. 

The initial data spec that @ingalls worked on, looked something like this: 

    {
        id: 3,
        properties: {
            'carmen:text': '9th st nw and 15th st nw,15th st nw and 9th st nw',
            'carmen:center': [0,0],
        },
        geometry: {
            type: 'Point',
            coordinates: [0,0]
        }
    };

The problem with this approach however was that because of the aforementioned limits we have in place, we only ever surface intersections and only when the user completely types out the street name like `9th Street Northwest` do we surface the actual street. The underlying problem being that we don’t have a good way to know whether the user is searching for an intersection or just the street. This PR aims to trigger intersection search only after the user starts typing `and`. 

What new approach are we using:

### Index Time:
1. At indexing time, we index intersections in the `carmen:intersections` in the following manner: 
`'carmen:intersections': [['highway number 2']]` → `['+intersection highway number 2 ,']`
We use `+intersection` as a special tag for intersections and we use the `,` as a delimiter between the intersection and the actual street name. 

2. Since, all features have the same score and relevance for addresses it is likely that intersections would get filtered out even when the user is searching for an intersection, to get around this we increase the weights for intersections to make sure that they are surfaced as results when someone is looking for an intersection. 
```
    if (weights[i] === -2) {
        weights[i] = 0.4;
    }
    else {
        weights[i] = Math.max(weights[i] / weightsum) * 0.6;
    }
```
tokens that have a `+intersection` are given a constant weight of 0.4 and the rest of the tokens are scaled between 0-0.6.

3. We standarize Multipoints to GeometryCollections to make it easier to generate intersection data and also reduce the number of features indexed/ size of the index. Therefore, now intersections don’t have to be a separate address cluster.

### Query Time:
4. At query time, we look for the keyword  ‘and’ which we use as a trigger for intersection search. For example, `'9th street northwest and f street northwest'` gets tokenized to `['+intersection', '9th', 'street', 'northwest', ',', 'f', 'street', 'northwest']`
5. We check whether the search is an intersection search by checking whether the cover text has the special tag `+intersection` if it does then we look up the corresponding geometry in the `GeometryCollection`. 
6. Once we have the geometry and the intersection name we stitch the place_name together which return to the user.

### Data Spec:
1. For the data design, let’s assume: 
- 9th Street Northwest which is also known as US Road, Highway Number 6
- F Street Northwest which is also known as Highway Number 4

#### Description: 

- 9th Street Northwest intersects with F Street Northwest at [0,2] and Highway Number 2 at [0,1]
- The aliases for 9th Street Northwest go in carmen:text separated by a comma. While the streets 9th Street Northwest intersects with goes in the carmen:intersections property.
- The coordinates in the geometry correspond to the point of intersection of the street in carmen:intersections and the feature in carmen:text.
    
```    
    Example data spec for intersections:
    {
        const address = {
            id:2,
            properties: {
                // Synonyms of the feature go in 'carmen:text'
                'carmen:text': 'US Road,Highway Number 6,9th Street Northwest',
                'carmen:center': [0,0],
                // intersections with the feature go here
                'carmen:intersections': ['F Street Northwest', 'Highway Number 2']
            },
            geometry: {
                type: 'MultiPoint',
                // coordinates correspond to the street in 'carmen:intersection'
                // for example, F Street Northwest and 9th Street Northwest intersect at [0,2]
                coordinates: [[0,2], [0,1]]
            }
        };
        queueFeature(conf.address, address, t.end);
    });
```

- Intersections can also represented as a GeometryCollection by adding  `carmen:intersections`: `[null,['1st Avenue', '2nd Avenue', '3rd Avenue'], null]` property
- And the corresponding coordinates in the geomteries property
-  The position in the geometries array corresponding to the non null array value in `'carmen:intersections'`, similar to addressnumbers
- We standardise a multipoint intersection feature as a GeometryCollection to allow intersections to be part of the same address cluster and avoid having to index a feature and it’s intersection separately

```
    {
        properties: {
            'carmen:text': 'Main Street Northwest',
            'carmen:addressnumber': [[1, 2, 3], null, null],
            'carmen:intersections': [null,['1st Avenue', '2nd Avenue', '3rd Avenue'], null],
            'carmen:rangetype': 'tiger',
            'carmen:parityl': [[], [], []],
            'carmen:parityr': [[], [], []],
            'carmen:lfromhn': [[], [], []],
            'carmen:rfromhn': [[], [], []],
            'carmen:ltohn': [[], [], []],
            'carmen:rtohn': [[], [], []]
        },
        geometry: {
            type: 'GeometryCollection',
            geometries: [{
                type: 'MultiPoint',
                coordinates: [[1,1], [2,2], [3,3]] <--- addressnumber MultiPoint array is 1st, since addressnumber: [[1, 2, 3], null, null]
            },{
                type: 'MultiPoint',
                coordinates: [[2,2], [3,2], [3,3]] <--- intersections MultiPoint array is 2nd, since intersections: [null,['1st Avenue', '2nd Avenue', '3rd Avenue'], null]
            },{
                type: 'MultiLineString',
                coordinates: [[[1,1], [2,2], [3,3]]]
            }]
        }
    };
```

- If there is more than one name for F Street Northwest and it intersects with 9th Street Northwest, the alias is added to carmen:intersections and the corresponding coordinates in the coordinates property of the geometry.

### Summary of Changes
- [ ] Adds unit tests for intersection search 
   - Does this new approach surface streets before intersections 
   - Are we able to find intersections?
   - Can we find intersections for abbreviated versions of 9th Street Northwest and F Street Northwest like 9th st nw and F st nw
- [ ] Adds special indexing intersections
- [ ] Adds geomtery lookup for intersections in the geometry collection
- [ ] Support to return the name of the street and the intersection
- [ ] New weighting for when the tokens have `+intersection`
- [ ] Don’t hard code the replacement `and`, have a way to specify this as a part of the index
- [ ] Trigger an intersection search only after the user types 'and'

Next Actions:
- [ ] Check with real index data generated by pt2itp cc @ingalls 
- [ ] Can this approach of tagged text we used beyond `and` for other languages or other query patterns such as `between X and Y`
- [ ] Global token support for allowing users to query for Street A & Street B

cc @mapbox/search

### CODE REVIEW

File: test/unit/text-processing/token.replaceToken.test.js [hunk 1]
```diff
@@ -314,3 +314,12 @@ test('replace complex global tokens', (t) => {
     t.deepEqual(token.replaceToken(replacer, 'talstr 3-5'), { query: 'tal str 3-5', lastWord: false }, 'talstr 3-5 => tal str 3-5');
     t.end();
 });
+
+test('replace complex global tokens', (t) => {
+    const replacer = token.createGlobalReplacer({
+        '&': 'and'
```
Comment (aarthykc) [inline test/unit/text-processing/token.replaceToken.test.js]: Adding this global token replacement causes `AT&T` to change to `ATandT`

File: lib/text-processing/termops.js [hunk 2]
```diff
@@ -573,6 +578,17 @@ function numTokenize(text, version) {
 
             numTokenized.push(replaced);
         }
+        if (replaced[i] === 'and') {
+            // f st +++ 9th st
+            replaced[i] = replaced[i].replace('and', '+++');
```
Comment (miccolis) [inline lib/text-processing/termops.js]: I think we can just drop this line, immediately below we remove the element from the array.

File: lib/text-processing/termops.js [hunk 3]
```diff
@@ -573,6 +578,17 @@ function numTokenize(text, version) {
 
             numTokenized.push(replaced);
         }
+        if (replaced[i] === 'and') {
```
Comment (miccolis) [inline lib/text-processing/termops.js]: One thing we should try and avoid is hardcoding any assumptions about english always being the language of our data. Perhaps we could consider using something that is explicitly a placeholder here.

File: lib/geocoder/verifymatch.js [hunk 4]
```diff
@@ -188,7 +192,7 @@ function verifyFeatures(query, geocoder, spatialmatches, loaded, options) {
             // and fell back to a street result. This penalty used to live in phrasematch
             // before being moved here to allow street level fallback mode
             if (feat.properties['carmen:address'] === false) cover.relev *= 0.99;
-
+            if (feat.intersections) cover.relev *= 0.99;
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: Did you end up finding that we still need this after moving to a prefix string based approach? ...or is this just a remnant of the earlier attempt?

File: lib/geocoder/verifymatch.js [hunk 5]
```diff
@@ -119,7 +119,10 @@ function verifyFeatures(query, geocoder, spatialmatches, loaded, options) {
 
         // Calculate to see if there is room for an address in the query based on bitmask
         // This recognizes 100 in the query list - `['1', 'fake', 'street', '100']` and returns the position of the addressnumber (3) and the addressnumber (100)
-        const address = termops.maskAddress(query, cover.text, cover.mask);
+        let address = termops.maskAddress(query, cover.text, cover.mask);
+        if (feats[0].intersections) address = {
+            addr: feats[0].properties['carmen:addressnumber'].slice(-1)[0]
+        };
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: It's not clear to me the purpose  of this address object editing is. What am I missing?

File: lib/text-processing/termops.js (hunk 3 above)
Comment (aarthykc) [reply lib/text-processing/termops.js]: Not done this yet, @miccolis, will add this at the end!

**Comment by ingalls** [comment]
I've included a bunch of new tests and some samples of the classes of intersection features

Ref: https://github.com/mapbox/carmen/pull/818

Basically it can have two variants
- Orphaned intersection - MultiPoint Feature => Geometry Collection
- Orphaned intersection Single Geometry Geometry Collection => No transformation needed
- Combined Intersection - GeometryCollection w/ Address => No transformation needed
- Combined Intersection - GeometryCollection w/ Network => No transformation needed
- Combined Intersection - GeometryCollection w/ Network & Address => No transformation needed

File: lib/geocoder/verifymatch.js [hunk 6]
```diff
@@ -120,9 +120,25 @@ function verifyFeatures(query, geocoder, spatialmatches, loaded, options) {
         // Calculate to see if there is room for an address in the query based on bitmask
         // This recognizes 100 in the query list - `['1', 'fake', 'street', '100']` and returns the position of the addressnumber (3) and the addressnumber (100)
         const address = termops.maskAddress(query, cover.text, cover.mask);
+        let intersection;
+        // cover.text for intersections will look like - +intersection f street northwest , 9th street
+        // i.e they will be indexed with +intersection
+        if (cover.text.indexOf('+intersection') > -1) {
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: Could this be more specific and test that `cover.text.indexOf('+intersection') === 0`? ...or are the circumstances where it wouldn't be the first bit of text?

File: lib/geocoder/verifymatch.js [hunk 7]
```diff
@@ -120,9 +120,25 @@ function verifyFeatures(query, geocoder, spatialmatches, loaded, options) {
...
+            intersection = cover.text.split(',')[0].replace('+intersection', '').trim();
+        }
 
         if (source.geocoder_address) {
-            if (address && (feats[0].properties['carmen:addressnumber'] || feats[0].properties['carmen:rangetype'])) {
+            if (intersection) {
+                const exactMatchIntersection = feats[0].properties['carmen:intersections'].find((f) => { return intersection === f.toLowerCase(); });
+                // Get the index of the intersection in the ['carmen:intersections'] property
+                // and then look up the coordinates in geometry coordinates of the feature
+                const intersectionCoordinates = feats[0].properties['carmen:intersections'].findIndex((f) => { return intersection === f.toLowerCase(); });
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: Could we avoid iterating over the `feats[0].properties['carmen:intersections']` array twice?

File: lib/geocoder/ops.js [hunk 8]
```diff
@@ -46,7 +46,9 @@ function getPlaceName(context, formatString, language, languageMode, matched) {
     let place_name;
     // Use geocoder_format to format output if applicable
     if (!formatString || formatString === true || formatString === 1) {
-        place_name = ((feat.properties['carmen:address'] ? feat.properties['carmen:address'] + ' ' : '') +
+        place_name = (
+            (feat.properties['carmen:intersection'] ? feat.properties['carmen:intersection'] + ' and ' : '') +
+            (feat.properties['carmen:address'] ? feat.properties['carmen:address'] + ' ' : '') +
```
Comment (miccolis) [inline lib/geocoder/ops.js]: This makes it seem like we'd support `carmen:intersection` and `carmen:address` on a single feature. Is that the case?

File: test/acceptance/geocode-unit.address-intersections.test.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        c.geocode('F Street Northwest and 9th Street Northwest', {}, (err, res) => {
+            t.ifError(err);
+            t.deepEquals(res.features[0].place_name, 'F Street Northwest and 9th Street Northwest', 'F Street Northwest and 9th Street Northwest');
+            t.deepEquals(res.features[0].geometry, { type: 'Point', coordinates: [0,2] }, 'Returns the correct geometry for F Street Northwest and 9th Street Northwest');
+            t.end();
+        });
+    });
+
+    tape('Searching for the intersection - F street northwest and 9th street northwest', (t) => {
+        c.geocode('F Street Northwest and', {}, (err, res) => {
```
Comment (miccolis) [inline test/acceptance/geocode-unit.address-intersections.test.js]: Just for completeness it would be nice to see a few more tests around the boundary, things like; `f st nw and`, `f st nw and 9`, `f st nw and 9th st`, `f street nw and 9th stre`

File: lib/geocoder/verifymatch.js [hunk 10]
```diff
@@ -308,7 +323,13 @@ function verifyContexts(contexts, sets, indexes, options, geocoder) {
                 cover = verify[ctx.properties['carmen:tmpid']];
                 ctx.properties['carmen:matches_language'] = cover.matches_language;
                 ctx.properties['carmen:prefix'] = cover.prefix;
-                ctx.properties['carmen:query_text'] = cover.text;
+                // for intersections we want the feature 9th street where the cover.text = +intersection f street northwest , 9th street
+                if (ctx.properties['carmen:intersection']) {
+                    // only if the name of the street has been typed out
+                    if (cover.text.indexOf(',') > -1) {
+                        ctx.properties['carmen:query_text'] = cover.text.split(',')[1].trim();
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: Something about this makes me a little nervous. Are there any cases where we can't assume that comma is going to be in the place we think it is? ...if it was ever missing we'd crash the process because element 1 would be undefined.

File: lib/text-processing/closest-lang.js [hunk 11]
```diff
@@ -319,6 +319,15 @@ function getLanguageCode(str) {
  */
 function getText(language, properties) {
     if (!properties['carmen:text']) throw new Error('Feature has no carmen:text');
+    // query_text for intersections use case is undefined when the user has not typed out the street after the and
+    if (properties['carmen:intersection'] && properties['carmen:query_text'] !== undefined) {
```
Comment (miccolis) [inline lib/text-processing/closest-lang.js]: To play devil's advocate; is this really the right place for code that tests against the user's query? The rest of the method is narrowly focused on the language and this addition seems to change the scope of the method.

**Comment by ingalls** [comment]
Working on supporting the new spec here: https://github.com/ingalls/pt2itp/pull/394

**Comment by aarthykc** [comment]
Added a better explanation about the approach in the OP.

File: lib/geocoder/ops.js [hunk 12]
```diff
@@ -76,8 +80,13 @@ function getPlaceName(context, formatString, language, languageMode, matched) {
                 // If we found a match between the template property and one within the returned context
                 if (templateProp === carmenProp) {
                     if (templateSubprop === '_name') {
+                        if (f.properties['carmen:intersection']) {
+                            const intersectionPrefix = feat.properties['carmen:intersection'] + ' and ';
+                            val = (intersectionPrefix + context.map((f) => { return (f.properties['carmen:intersection'] ? getIntersectionName(geocoder, f.properties) : undefined); }));
```
Comment (miccolis) [inline lib/geocoder/ops.js]: If I follow this correctly it's rely on the Array to string case to turn this into a comma separated list... but won't any undefined elements end up as ",," in the resultant string?

File: lib/geocoder/ops.js [hunk 13]
```diff
@@ -41,12 +41,16 @@ function getFormatString(context, format, language) {
  * @param {Boolean} matched If set, use matched text in place name
  * @return {String} formatted place_name string
  */
-function getPlaceName(context, formatString, language, languageMode, matched) {
+function getPlaceName(context, formatString, language, languageMode, matched, geocoder) {
```
Comment (miccolis) [inline lib/geocoder/ops.js]: The js doc isn't updated, but if I follow the 'geocoder' param is the geocoder itself. I'd expect that to be the first param.

File: lib/geocoder/ops.js [hunk 14]
```diff
@@ -321,3 +335,24 @@ function getMatchingText(item, geocoder, requestedLanguage) {
         return matched;
     }
 }
+
+/*
+* getIntersectionName - helper function to get the street to add to the cross street
+* @param {Object} geocoder Geocoder object
+* @param {object} properties A feature properties object
```
Comment (miccolis) [inline lib/geocoder/ops.js]: Can we get a '@return' doc here. thx.

File: lib/geocoder/verifymatch.js [hunk 15]
```diff
@@ -121,8 +121,41 @@ function verifyFeatures(query, geocoder, spatialmatches, loaded, options) {
...
+
         if (source.geocoder_address) {
-            if (address && (feats[0].properties['carmen:addressnumber'] || feats[0].properties['carmen:rangetype'])) {
+            if (intersection) {
+                let exactMatchIntersection, intersectionCoordinates, geometryPos;
+                // Get the index of the intersection in the ['carmen:intersections'] property
+                // and then look up the coordinates in geometry coordinates of the feature
+                const intersectionsArray = feats[0].properties['carmen:intersections'];
+                for (let i = 0; i < intersectionsArray.length; i++) {
+                    if (!intersectionsArray[i]) continue;
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: When would we have an empty element in this array?

File: lib/geocoder/verifymatch.js [hunk 16]
```diff
@@ -121,8 +121,41 @@ function verifyFeatures(query, geocoder, spatialmatches, loaded, options) {
...
+                    // geometryPos is position of the intersection multipoint array in the GeometryCollection
+                    geometryPos = intersectionsArray.indexOf(intersectionsArray[i]);
+                    for (let j = 0; j < intersectionsArray[i].length; j++) {
+                        const indexedIntersectionText = termops.getIndexableText(source.simple_replacer, source.complex_query_replacer, [],
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: nit pick, but can we change this variable name, it makes me think we're about do index something.

Also, I'm not sure of a better way, but getIndexableText is a pretty heavy function. Are they any alternatives?

File: lib/text-processing/termops.js [hunk 17]
```diff
@@ -369,13 +390,20 @@ function getMinimalIndexableText(replacer, globalTokens, doc) {
...
         if (!tokens.length) continue;
 
-        if (housenumRange) {
+        if (intersections) {
+            let l = intersections.length;
+            while (l--) {
+                const withIntersections = [intersections[l]].concat(tokens);
+                add(withIntersections);
+            }
+        } if (housenumRange) {
```
Comment (miccolis) [inline lib/text-processing/termops.js]: This `if` statement should be on a new line.

File: lib/text-processing/termops.js [hunk 18]
```diff
@@ -536,16 +571,28 @@ function getWeights(tokens, freq) {
...
                 weights[i] = 0.2;
-            } else {
+            }
+            else {
                 weights[i] = Math.max(weights[i] / weightsum) * 0.8;
             }
         }
-    } else {
+    }
+    // if the token has a +intersection give a constant weight of 0.4
```
Comment (miccolis) [inline lib/text-processing/termops.js]: I'm curious, how did you pick 0.4?

File: lib/geocoder/verifymatch.js (hunk 16 above)
Comment (aarthykc) [reply lib/geocoder/verifymatch.js]: I changed this to `tokenizedIntersectionText`, does that work better? I've added comments about why we need to do this. 

I agree, I see that portion of the getIndexableText where we tokenize a string based on the simple token replacement being helpful in other places as well. Are you saying we should move that to a separate function?

File: lib/text-processing/termops.js (hunk 17 above)
Comment (aarthykc) [reply lib/text-processing/termops.js]: Done!

File: lib/geocoder/verifymatch.js (hunk 15 above)
Comment (aarthykc) [reply lib/geocoder/verifymatch.js]: I've mentioned this in the data spec, checkout -- https://github.com/mapbox/carmen/pull/808/files#diff-1dacf7eebaabb0846d2e11bcdb3a6be2R40

File: lib/geocoder/ops.js (hunk 12 above)
Comment (aarthykc) [reply lib/geocoder/ops.js]: Yess, totally missed this which ended up ruining the formatting, I've fixed this in the next commit.

Comment (aarthykc) [reply lib/geocoder/ops.js]: `$(yarn bin)/carmen --config dev/mapbox.places.js --query "O Street Northwest and Naylor Court Northwest" --types address`

tokens
------
o, street, northwest, and, naylor, court, northwest

Features
--------
- 1.00 O Street Northwest and Naylor Court Northwest, Washington, District of Columbia 20001, United States (address.4561921048832130)

File: lib/geocoder/ops.js (hunk 14 above)
Comment (aarthykc) [reply lib/geocoder/ops.js]: done!

File: lib/text-processing/termops.js (hunk 18 above)
Comment (aarthykc) [reply lib/text-processing/termops.js]: This was a little bit of a trial, I tried with 0.2 but it was not high enough to bump intersections and so I ended up going with 0.4

File: lib/text-processing/termops.js [hunk 19]
```diff
@@ -620,3 +686,32 @@ function decode3BitLogScale(num, max) {
...
+            ranges.push(numToken);
+        }
+    }
+    if (doc.properties['carmen:intersections']) {
+        // carmen:intersections will look something like [null, [intersections array], null]
+        const keys = doc.properties['carmen:intersections'].filter(Boolean);
+        for (let i = 0; i < keys.length; i++) {
+            for (let j = 0; j < keys[i].length; j++) {
+                // indexes F Street Northwest as 'intersectionsearch <key>'
+                add('intersectionsearch ' + keys[i][j]);
```
Comment (aarthykc) [inline lib/text-processing/termops.js]: Since, the + in +intersection gets stripped anyway, chose to go with `intersectionsearch` and covert tokens to '+intersection' <streetname> ',' finally in tokenize()

File: test/acceptance/geocode-unit.address-intersections.test.js [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+
+    tape('Searching for the intersection - F st nw and 9th st', (t) => {
+        c.geocode('F st nw and 9th st', {}, (err, res) => {
+            t.deepEquals(res.features[0].place_name, 'F Street Northwest and 9th Street Northwest', 'F st nw and 9th st');
+            t.deepEquals(res.features[0].geometry, { type: 'Point', coordinates: [0,2] }, 'Returns the correct geometry for F Street Northwest and 9th Street Northwest');
+            t.end();
+        });
+    });
+
+    tape('Searching for the intersection - F st nw and 9th (should favour returning the street over the intersection)', (t) => {
```
Comment (aarthykc) [inline test/acceptance/geocode-unit.address-intersections.test.js]: This is because of the penalty I added in for intersections when we don't have enough information about the cross street, I feel fine returning an intersection only when we have more information instead of guessing.

File: lib/text-processing/termops.js (hunk 3 above)
Comment (aarthykc) [reply lib/text-processing/termops.js]: done

File: lib/geocoder/ops.js (hunk 13 above)
Comment (miccolis) [reply lib/geocoder/ops.js]: Ping on this again. I really think this function signature needs to be changed, or maybe it's too overloaded and needs to be refactored? I'm not sure.

File: lib/geocoder/ops.js [hunk 21]
```diff
@@ -321,3 +336,25 @@ function getMatchingText(item, geocoder, requestedLanguage) {
...
+* @return String returns the correct carmen:text synonym based on the input query
+*/
+function getIntersectionStreetName(geocoder, properties) {
```
Comment (miccolis) [inline lib/geocoder/ops.js]: A few things about this method kinda bother me. At a high level I think I just can't see all the implications of the aggressive string manipulations here. Should we have unit tests of this method?

File: lib/geocoder/verifymatch.js [hunk 22]
```diff
@@ -121,8 +121,53 @@ function verifyFeatures(query, geocoder, spatialmatches, loaded, options) {
...
+                    if (!intersectionsArray[i]) continue;
+                    // geometryPos is position of the intersection multipoint array in the GeometryCollection
+                    geometryPos = intersectionsArray.indexOf(intersectionsArray[i]);
+                    for (let j = 0; j < intersectionsArray[i].length; j++) {
+                        // while indexing we tokenize the intersection array
+                        // to find the exact match intersection we have to tokenize it first before we attempt to compare them
+                        const tokenizedIntersectionText = termops.getIndexableText(source.simple_replacer, source.complex_query_replacer, [],
+                            { properties: { 'carmen:text': intersectionsArray[i][j] } }
+                        );
+                        if (intersection === tokenizedIntersectionText[0].tokens.join(' ')) {
```
Comment (miccolis) [inline lib/geocoder/verifymatch.js]: I'm not sure I follow the intention here. Won't getIndexableText generate a lot of options? Why would we do that and only look at the first?

File: lib/text-processing/termops.js [hunk 23]
```diff
@@ -126,6 +128,13 @@ function tokenize(query, lonlat) {
         }
     }
 
+    // we check if the tokens have 'intersectionsearch' if so we know that it is intersection data
+    // add +intersection to the front of the array
+    // add a comma to the end of the array
+    if (tokens.includes('intersectionsearch')) {
```
Comment (miccolis) [inline lib/text-processing/termops.js]: What I liked about `+intersection` was that is used a reserved character that a user couldn't enter. Any way we can  prevent the "+" from getting stripped away?

More importantly, I don't follow why we'd want to have re-ordering happen in tokenize. Seems like that changes the nature of the method. Is there a reason we can't handle it all elsewhere?

File: lib/text-processing/termops.js (hunk 19 above)
Comment (miccolis) [reply lib/text-processing/termops.js]: Yea, this make me sad. Do you know where it's getting stripped?

File: lib/text-processing/termops.js [hunk 24]
```diff
@@ -560,7 +618,7 @@ function isAddressNumber(tokens) {
 }
 
 module.exports.numTokenize = numTokenize;
-function numTokenize(text, version) {
+function numTokenize(text, version, intersectionToken) {
```
Comment (miccolis) [inline lib/text-processing/termops.js]: Seems like we should change the name of the method if we're going to change it's scope here - AddressTokenize? ...even then this isn't about tokenization. variants?

File: test/unit/util/feature.addrTransform.test.js [hunk 25]
```diff
@@ -0,0 +1,388 @@
...
+});
+
+tape('Combined Features', (t) => {
+    t.deepEquals(transform({
+        type: 'Feature',
+        properties: {
+            'carmen:text': 'Main Street Northwest',
+            'carmen:addressnumber': [null, [1, 2, 3]],
+            'carmen:rangetype': 'tiger',
+            'carmen:parityl': [[], null],
```
Comment (miccolis) [inline test/unit/util/feature.addrTransform.test.js]: I'm curious why are these sometimes null and sometimes the empty array? Is there a meaningful difference? Should we prefer one form? ...personally I like empty array's better

File: lib/text-processing/termops.js (hunk 19 above)
Comment (aarthykc) [reply lib/text-processing/termops.js]: This is getting stripped in tokenize, which is why i tag the `+intersection` and add the comma there given that both get stripped earlier on in the tokenize function

File: lib/geocoder/ops.js (hunk 21 above)
Comment (aarthykc) [reply lib/geocoder/ops.js]: sure, I can add unit tests

File: lib/text-processing/termops.js [hunk 26]
```diff
@@ -573,6 +633,14 @@ function numTokenize(text, version) {
 
             numTokenized.push(replaced);
         }
+        if (!!intersectionToken && replaced[i] === intersectionToken) {
```
Comment (miccolis) [inline lib/text-processing/termops.js]: @aarthykc as per chat - what I'm wondering is if we should make sure that the token we're examining isn't the last one. A possible check could be `... && i  < text.length - 1`. that way we'd avoid looking for cross streets when we don't have any info about one yet.

**Comment by aarthykc** [comment]
Joint forces with @miccolis post tokenization refactor to create a new PR - #838
